src/main/java/org/terracotta/build/services/BuildInfo.java

```diff
@@ -0,0 +1,148 @@
+package org.terracotta.build.services;
+
+import org.gradle.api.Project;
+import org.gradle.api.provider.Property;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+
+import java.io.Serializable;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.List;
+import java.util.Objects;
+import java.util.regex.Pattern;
+
+import static java.lang.Integer.parseInt;
+import static java.util.Arrays.asList;
+import static java.util.stream.Collectors.joining;
+
+public abstract class BuildInfo implements BuildService<BuildInfo.Parameters> {
```

Comment:
For 1)

I didn't want a user extension which can be configured: I wanted a service that provides information about the build in the immutable manner.

I went to the fastest route to expose directly the BuildInfo object instead of making an interface, but I can make an interface.

For 2)

We do not force that: the choice of the scope if decided how the shared service is constructed and used. I just provided a singleton method to get a default one.

```
  public static BuildInfo getOrInstall(Project project) {
    return project.getGradle().getSharedServices().registerIfAbsent("buildInfo", BuildInfo.class, spec -> spec.parameters(parameters -> {
      parameters.getDefaultProjectVersion().set(project.findProperty("defaultVersion"));
      parameters.getOverrideProjectVersion().set(project.findProperty("overrideVersion"));
      parameters.getProjectVersion().set(project.getVersion());
      parameters.getGit().set(Git.getOrInstall(project));
    })).get();
  }
```

But nothing prevents another plugin or task to use a newly created service that would be constructed like this:

```
return project.getGradle().getSharedServices().registerIfAbsent("myOtherPorjectBuildInfo", BuildInfo.class, spec ->
```